### PR TITLE
[MIDDLEWARE] Ot198 26 - Middleware for Authenticated Users

### PR DESCRIPTION
**Pull request by maxisosa89**
[OT198-26](https://alkemy-labs.atlassian.net/browse/OT198-26)

## Description:

AS authenticated user
I WANT to restrict access to certain content
TO keep data protected

### Criteria of acceptance:

It should check if the request contains the token corresponding to a user, otherwise it should return an error code 403 with an error message.

### Evidence:

image

### CODE REVIEW

File: middlewares/auth.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const authConfig = require('../config/auth')
+const { User } = require('../database/models')
+
+module.exports = (req, res, next) => {
+  try {
+    if (!req.headers.authorization) {
+      res.status(403).json({ msg: 'Access not authorized.' })
+    } else {
+      const token = req.headers.authorization.split(' ')[1]
+      jwt.verify(token, authConfig.secret, async (err, decoded) => {
```
Comment (gabodmc) [inline middlewares/auth.js:12]: Aca hay que separar las responsabilidades con otro middleware (ej jwt.js) que se encargue de manejar la generacion y verificacion de tokens

**Review by gabodmc — CHANGES_REQUESTED** [review]
Muy buen PR Maxi, solo un pequeño detalle

File: middlewares/auth.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+const createHttpError = require('http-errors')
+const jwt = require('jsonwebtoken')
+const authConfig = require('../config/auth')
+const { User } = require('../database/models')
+
+module.exports = (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (maxisosa89) [reply middlewares/auth.js]: Agregado

File: middlewares/jwt.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+const jwt = require('jsonwebtoken')
+const { User } = require('../database/models')
+const authConfig = require('../config/auth')
+
+module.exports = {
+  validateToken: (token) => (req, res, next) => {
+    jwt.verify(token, authConfig.secret, async (err, decoded) => {
+      if (err) {
+        res.status(403).json({ msg: 'Token not decoded' })
```
Comment (honeybadger2788) [inline middlewares/jwt.js:9]: Idem auth middleware

**Review by honeybadger2788 — APPROVED** [review]
Excelente, Max! Solo unos comentarios sobre el response

File: middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const createHttpError = require('http-errors')
+const { validateToken } = require('./jwt')
+
+module.exports = {
+  auth: (req, res, next) => {
+    try {
+      if (!req.headers.authorization) {
+        res.status(403).json({ msg: 'Access not authorized.' })
```
[the remark this answers is not in the document]
Comment (maxisosa89) [reply middlewares/auth.js]: Se aplica el helper

**Review by gabodmc — APPROVED** [review]

**Review by nicobes14 — APPROVED** [review]
Buenisimo Maxi!

**Review by HernanGarcia1985 — APPROVED** [review]
Todo OK!!!

**Pull request merged**